### PR TITLE
Made initialization of self.buildMetadataIdentifier safer and cleaner

### PR DESCRIPTION
**Pull request by rsmoz**
In the words of @CodaFi: "Collapsed the Pyramid of Doom"

### CODE REVIEW

**Comment by rsmoz** [comment]
In the words of @CodaFi: "Collapsed the Pyramid of Doom"

**Comment by rsmoz** [comment]
I also have [this](https://github.com/rsmoz/swift-package-manager/commit/1e3c8abf7b2723eac34da980a63806efa981e8ff) which I can add to this PR (after a squash) but thought that it might have taken the conciseness too far. This, unlike the earlier change, doesn't really enhance safety. But it's here if you find it interesting.

**Comment by rsmoz** [comment]
I also have [this](https://github.com/rsmoz/swift-package-manager/commit/1e3c8abf7b2723eac34da980a63806efa981e8ff) which I can add to this PR (after a squash) but thought that it might have taken the conciseness too far. This, unlike the earlier change, doesn't really enhance safety. But it's here if you find it interesting.

**Comment by kostiakoval** [comment]
I agree that nested `if let` and `Pyramid of Doom` is not a good thing but. I also think that the previous code is more readable and clear what it does. 
Clever code is harder to understand, so I would keep the old version, maybe refactor it just a little bit

**Comment by rsmoz** [comment]
@kostiakoval My issue here isn't even with the style of the code, just that its [statefulness](https://github.com/rsmoz/swift-package-manager/commit/a0f2fbd1b9b95571c31cae6316cde779e167e60f#diff-2c6beb1fad4deea16628c7a8c329d4d8L55) is more likely to lead to bugs.

But I still think that my PR _is_ more readable. The pyramid of doom is harder to parse; I have to figure out which variables/constants are being used where, and I have to differentiate very similar names. Have a short series of concise maps is much clearer to me at a glance. As for refactoring the original... I don't really see a better way than the map/flatMap approach. Those functions are there for a reason.

**Comment by rsmoz** [comment]
For instance, I know what [this](https://github.com/rsmoz/swift-package-manager/commit/a0f2fbd1b9b95571c31cae6316cde779e167e60f#diff-2c6beb1fad4deea16628c7a8c329d4d8L56) means but when all these variable names are mixed together, it becomes harder to find what I'm looking for

File: Sources/PackageDescription/Version.swift [hunk 1]
```diff
@@ -52,14 +52,12 @@ public struct Version {
...
-        if let metadataStartIndex = metadataStartIndex {
-            let buildMetadataCharacters = characters.suffixFrom(metadataStartIndex.successor())
-            if !buildMetadataCharacters.isEmpty {
-                buildMetadataIdentifier = String(buildMetadataCharacters)
-            }
-        }
-        self.buildMetadataIdentifier = buildMetadataIdentifier
+        self.buildMetadataIdentifier = metadataStartIndex
+                                       .map { characters.suffixFrom($0.successor()) }
+                                       .flatMap { $0.isEmpty ? nil : $0 }
```
Comment (mxcl) [inline Sources/PackageDescription/Version.swift]: `filter`?

File: Sources/PackageDescription/Version.swift [hunk 2]
```diff
@@ -52,14 +52,12 @@ public struct Version {
...
+                                       .map { String($0) }
```
Comment (mxcl) [inline Sources/PackageDescription/Version.swift]: `map(String.init)`?

**Comment by mxcl** [comment]
Yeah, I don't really feel this is more readable. Even though generally I love this style.

File: Sources/PackageDescription/Version.swift (hunk 1 above)
Comment (CodaFi) [inline Sources/PackageDescription/Version.swift]: Needs `mfilter`, unfortunately.

**Comment by rsmoz** [comment]
But... but safety!

File: Sources/PackageDescription/Version.swift (hunk 2 above)
Comment (rsmoz) [inline Sources/PackageDescription/Version.swift]: I tried, the compiler doesn't like it.

**Comment by mxcl** [comment]
Unless I'm missing something, it isn't any safer.

File: Sources/PackageDescription/Version.swift (hunk 2 above)
Comment (mxcl) [inline Sources/PackageDescription/Version.swift]: Right, probably `String(characters: …)` has a default parameter.

**Comment by rsmoz** [comment]
You're right, "safer" probably isn't the way to describe it. But with the original version, if by chance you happened to accidentally delete

```
if let metadataStartIndex = metadataStartIndex {
            let buildMetadataCharacters = characters.suffixFrom(metadataStartIndex.successor())
            if !buildMetadataCharacters.isEmpty {
                buildMetadataIdentifier = String(buildMetadataCharacters)
            }
}
```

it would still compile, whereas accidentally deleting 

```
self.buildMetadataIdentifier = metadataStartIndex
                                       .map { characters.suffixFrom($0.successor()) }
                                       .flatMap { $0.isEmpty ? nil : $0 }
                                       .map { String($0) }
```

would be caught by the compiler.

**Comment by rsmoz** [comment]
That being said, I can understand why some may not consider the map approach as readable as the current implementation. Closing the PR.

**Comment by mxcl** [comment]
True, we could certainly improve this code

**Pull request closed without merging**